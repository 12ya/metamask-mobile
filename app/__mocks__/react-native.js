import { Linking, DeviceEventEmitter } from 'react-native';

Linking.addEventListener = jest.fn();
Linking.removeEventListener = jest.fn((event, callback) => {
  // Simulate the behavior of removeEventListener
  if (typeof callback === 'function') {
    callback();
  }
});
Linking.openURL = jest.fn();
Linking.canOpenURL = jest.fn();
Linking.getInitialURL = jest.fn();

DeviceEventEmitter.addListener = jest.fn();
DeviceEventEmitter.removeListener = jest.fn((event, callback) => {
  // Simulate the behavior of removeListener
  if (typeof callback === 'function') {
    callback();
  }
});

export default { Linking, DeviceEventEmitter };
