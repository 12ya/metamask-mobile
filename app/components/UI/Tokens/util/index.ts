export { handleBalance } from './handleBalance';
export { sortAssets } from './sortAssets';
