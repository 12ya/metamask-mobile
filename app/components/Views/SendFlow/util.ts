import { toChecksumAddress } from 'ethereumjs-util';
import {useSelector} from 'react-redux';
import {
    selectNetwork,
    selectChainId
  } from '../../../selectors/networkController';

  import {
    validateAddressOrENS,
  } from '../../../util/address';

const identities = useSelector(
    (state: any) =>
      state.engine.backgroundState.PreferencesController.identities,
  );

  const addressBook = useSelector(
    (state: any) =>
    state.engine.backgroundState.AddressBookController.addressBook
    );

    const network = useSelector(selectNetwork);
    const chainId = useSelector(selectChainId);

const getAddressNameFromBookOrIdentities = (toAccount) => {
    if (!toAccount) return;

    const networkAddressBook = addressBook[network] || {};

    const checksummedAddress = toChecksumAddress(toAccount);

    return networkAddressBook[checksummedAddress]
      ? networkAddressBook[checksummedAddress].name
      : identities[checksummedAddress]
      ? identities[checksummedAddress].name
      : null;
  };

    /**
   * This set to the state all the information
   *  that come from validating an ENS or address
   * @param {*} toSelectedAddress - The address or the ens writted on the destination input
   */
  export const validateAddressOrENSFromInput = async ({toAccount}) => {
    console.log(toAccount, 'tosAccount');
    
    const {
      addressError,
      toEnsName,
      addressReady,
      toEnsAddress,
      addToAddressToAddressBook,
      toAddressName,
      errorContinue,
      isOnlyWarning,
      confusableCollection,
    } = await validateAddressOrENS({
      toAccount,
      network,
      addressBook,
      identities,
      chainId,
    });

    return null

    // return {addressError, toEnsName, toSelectedAddressReady: addressReady, toEnsAddressResolved: toEnsAddress, addToAddressToAddressBook, toSelectedAddressName: toAddressName, errorContinue, isOnlyWarning, confusableCollection}
  };

// export const onToSelectedAddressChange = ({toAccount, updateParentState}) => {

//     const addressName = getAddressNameFromBookOrIdentities(toAccount);
    
//     /**
//      * If the address is from addressBook or identities
//      * then validation is not necessary since it was already validated
//      */
//     if (addressName) {
//         updateParentState({toAccount, toSelectedAddressReady: true, isFromAddressBook: true, toSelectedAddressName: addressName})
//     } else {
//       validateAddressOrENSFromInput(toAccount);
//       /**
//        * Because validateAddressOrENSFromInput is an asynchronous function
//        * we are setting the state here synchronously, so it does not block the UI
//        * */
      
//       updateParentState({toAccount, isFromAddressBook: false, toSelectedAddressName: addressName, toSelectedAddressReady: false})
//     }
//   }; 