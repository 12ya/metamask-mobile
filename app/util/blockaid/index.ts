import { FlagType } from 'app/components/UI/BlockaidBanner/BlockaidBanner.types';

// eslint-disable-next-line import/prefer-default-export
export const showBlockaidUI = () =>
  process.env.MM_BLOCKAID_UI_ENABLED === 'true';

export const getAdditionalMetricsParams = (transaction: any) => {
  const { result_type, reason } = transaction.securityAlertResponse;
  let uiCustomizations;
  
  if (result_type === FlagType.Malicious) {
    uiCustomizations = ['flagged_as_malicious'];
  }

  const additionalParams: Record<string, any> = {
    ui_customizations: uiCustomizations,
  };

  if (result_type !== FlagType.Benign) {
    additionalParams['security_alert_response'] = result_type;
    additionalParams['security_alert_reason'] = reason;
  }

  return additionalParams;
};
