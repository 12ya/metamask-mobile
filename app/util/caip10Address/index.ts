import { isValidHexAddress } from '../address';

// Types for parsed CAIP-10 address components
export interface Caip10Address {
  accountId: string;
  chainId: string;
  address: string;
}

/**
 * Parses a CAIP-10 formatted address into its components.
 *
 * CAIP-10 addresses follow the format `namespace:reference:address`,
 * where `namespace` and `reference` identify the blockchain, and `address`
 * specifies the account on that blockchain. This function extracts and returns
 * these components as an object. It throws an error for invalid formats.
 *
 * @param {string} caip10Address - The CAIP-10 formatted address to parse.
 * @returns {Caip10Address} An object containing the parsed `accountId`, `chainId`,                          and `address`.
 * @throws {Error} If `caip10Address` does not conform to the CAIP-10 format.
 *
 * @example
 * parseCaip10Address('eip155:1:0x1234567890abcdefABCDEF1234567890ABCDEF');
 * // Returns: { accountId: 'eip155', chainId: '1', address: '0x1234567890abcdefABCDEF1234567890ABCDEF' }
 */
export function parseCaip10Address(caip10Address: string): Caip10Address {
  if (/\s/.test(caip10Address)) {
    throw new Error('CAIP-10 address format cannot contain spaces');
  }

  const parts = caip10Address.split(':');
  if (parts.length !== 3) {
    throw new Error('Invalid CAIP-10 address format');
  }
  const [accountId, chainId, address] = parts;

  // For now we only support EVM addresses, so we only validate those
  if (accountId === 'eip155') {
    if (!isValidHexAddress(address)) {
      throw new Error('Invalid Ethereum address format');
    }
  } else {
    throw new Error('Only support Ethereum addresses at the moment');
  }

  return { accountId, chainId, address };
}
