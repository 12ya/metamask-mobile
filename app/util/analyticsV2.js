import { InteractionManager } from 'react-native';
import DefaultPreference from 'react-native-default-preference';
import Analytics from '../core/Analytics/Analytics';
import Logger from './Logger';
import { DENIED, METRICS_OPT_IN } from '../constants/storage';

/**
 * This takes params with the following structure:
 * { foo : 'this is not anonymous', bar: {value: 'this is anonymous', anonymous: true} }
 * @param {Object} eventName
 * @param {Object} params
 */
export const trackEventV2 = (eventName, params) => {
  const init = async () => {
    const metricsOptIn = await DefaultPreference.get(METRICS_OPT_IN);
    if (metricsOptIn === DENIED) return;

    InteractionManager.runAfterInteractions(() => {
      let anonymousEvent = false;
      try {
        if (!params || Object.keys(params).length === 0) {
          Analytics.trackEvent(eventName);
        }

        const userParams = {};
        const anonymousParams = {};

        for (const key in params) {
          const property = params[key];

          if (
            property &&
            typeof property === 'object' &&
            !Array.isArray(property)
          ) {
            if (property.anonymous) {
              anonymousEvent = true;
              // Anonymous property - add only to anonymous params
              anonymousParams[key] = property.value;
            } else {
              // Non-anonymous property - add to both
              userParams[key] = property.value;
              anonymousParams[key] = property.value;
            }
          } else {
            // Non-anonymous properties - add to both
            userParams[key] = property;
            anonymousParams[key] = property;
          }
        }

        // Log all non-anonymous properties
        if (Object.keys(userParams).length) {
          Analytics.trackEventWithParameters(eventName, userParams);
        }

        // Log all anonymous properties
        if (anonymousEvent && Object.keys(anonymousParams).length) {
          Analytics.trackEventWithParameters(eventName, anonymousParams, true);
        }
      } catch (error) {
        Logger.error(error, 'Error logging analytics');
      }
    });
  };
  init();
};

export default {
  trackEvent: trackEventV2,
};
