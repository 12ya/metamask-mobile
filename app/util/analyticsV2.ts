import { InteractionManager } from 'react-native';
import { MetaMetrics, IMetaMetricsEvent } from '../core/Analytics';
import Logger from './Logger';

/**
 * This takes params with the following structure:
 * { foo : 'this is not anonymous', bar: {value: 'this is anonymous', anonymous: true} }
 * @param {Object} eventName
 * @param {Object} params
 */
const trackEventV2 = (event: IMetaMetricsEvent, params: any) => {
  InteractionManager.runAfterInteractions(() => {
    let anonymousEvent = false;
    try {
      if (!params && !event.properties) {
        MetaMetrics.trackEvent(event.name, {});
      }

      const userParams = { ...event.properties } as any;
      const anonymousParams = {} as any;

      for (const key in params) {
        const property = params[key];

        if (
          property &&
          typeof property === 'object' &&
          !Array.isArray(property)
        ) {
          if (property.anonymous) {
            anonymousEvent = true;
            // Anonymous property - add only to anonymous params
            anonymousParams[key] = property.value;
          } else {
            // Non-anonymous property - add to both
            userParams[key] = property.value;
            anonymousParams[key] = property.value;
          }
        } else {
          // Non-anonymous properties - add to both
          userParams[key] = property;
          anonymousParams[key] = property;
        }
      }

      // Log all non-anonymous properties
      if (Object.keys(userParams).length) {
        MetaMetrics.trackEvent(event.name, userParams);
      }

      // Log all anonymous properties
      if (anonymousEvent && Object.keys(anonymousParams).length) {
        MetaMetrics.trackAnonymousEvent(event.name, anonymousParams);
      }
    } catch (error: any) {
      Logger.error(error, 'Error logging analytics');
    }
  });
};

/**
 * This functions logs errors to analytics instead of sentry.
 * The objective is to log errors (that are not errors from our side) like “Invalid Password”.
 * An error like this generally means a user inserted the wrong password, so logging to sentry doesn't make sense.
 * But we still want to log this to analytics so that we are aware of a rapid increase which may mean it's an error from our side, for example, an error with the encryption library.
 * @param {String} type
 * @param {String} errorMessage
 * @param {String} otherInfo
 */
const trackErrorAsAnalytics = (
  type: string,
  errorMessage: string,
  otherInfo: string,
) => {
  try {
    MetaMetrics.trackEvent('Error occurred', {
      error: true,
      type,
      errorMessage,
      otherInfo,
    });
  } catch (error: any) {
    Logger.error(error, 'Error logging analytics - trackErrorAsAnalytics');
  }
};

export default {
  trackEvent: trackEventV2,
  trackErrorAsAnalytics,
};
