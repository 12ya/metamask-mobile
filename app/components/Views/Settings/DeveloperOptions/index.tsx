import React, { useEffect } from 'react';
import { ScrollView } from 'react-native';

import { strings } from '../../../../../locales/i18n';
import { useTheme } from '../../../../util/theme';
import { getNavigationOptionsTitle } from '../../../UI/Navbar';
import { Props } from './DeveloperOptions.types';
import createStyles from './DeveloperOptions.styles';
import SentryTest from './SentryTest';

const DeveloperOptions = ({ navigation, route }: Props) => {
  const isFullScreenModal = route?.params?.isFullScreenModal;

  const theme = useTheme();
  const { colors } = theme;
  const styles = createStyles(colors);

  useEffect(
    () => {
      navigation.setOptions(
        getNavigationOptionsTitle(
          strings('app_settings.developer_options.title'),
          navigation,
          isFullScreenModal,
          colors,
          null,
        ),
      );
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [colors],
  );

  return (
    <ScrollView style={styles.wrapper}>
      <SentryTest />
    </ScrollView>
  );
};

export default DeveloperOptions;
