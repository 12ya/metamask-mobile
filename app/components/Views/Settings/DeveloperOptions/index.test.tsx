import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { render } from '@testing-library/react-native';

import { backgroundState } from '../../../../util/test/initial-root-state';
import { mockTheme, ThemeContext } from '../../../../util/theme';
import DeveloperOptions from './';

const mockStore = configureMockStore();

const initialState = {
  DeveloperOptions: {},
  engine: {
    backgroundState,
  },
};

const store = mockStore(initialState);

const setOptions = jest.fn();

describe('DeveloperOptions', () => {
  it('should render correctly', () => {
    const wrapper = render(
      <Provider store={store}>
        <ThemeContext.Provider value={mockTheme}>
          <DeveloperOptions
            navigation={{
              setOptions,
            }}
            route={{}}
          />
          ,
        </ThemeContext.Provider>
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
