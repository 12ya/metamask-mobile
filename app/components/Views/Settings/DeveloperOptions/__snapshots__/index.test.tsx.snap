// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeveloperOptions should render correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "#ffffff",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "elevation": 0,
                  "flex": 1,
                  "shadowColor": "transparent",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              >
                <TouchableOpacity
                  accessible={true}
                  activeOpacity={1}
                  disabled={false}
                  onPress={[Function]}
                  onPressIn={[Function]}
                  onPressOut={[Function]}
                  style={
                    {
                      "alignItems": "center",
                      "borderRadius": 8,
                      "height": 32,
                      "justifyContent": "center",
                      "marginHorizontal": 16,
                      "opacity": 1,
                      "width": 32,
                    }
                  }
                  testID="back-arrow-button"
                >
                  <SvgMock
                    color="#141618"
                    height={24}
                    name="ArrowLeft"
                    style={
                      {
                        "height": 24,
                        "width": 24,
                      }
                    }
                    width={24}
                  />
                </TouchableOpacity>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 590,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "#0376c9",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  Developer Options
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            collapsable={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTScrollView
                        style={
                          {
                            "backgroundColor": "#ffffff",
                            "flex": 1,
                            "padding": 24,
                            "paddingBottom": 48,
                          }
                        }
                      >
                        <View>
                          <Text
                            accessibilityRole="text"
                            style={
                              {
                                "color": "#141618",
                                "fontFamily": "EuclidCircularB-Bold",
                                "fontSize": 24,
                                "fontWeight": "700",
                                "letterSpacing": 0,
                                "lineHeight": 32,
                                "marginTop": 16,
                              }
                            }
                          >
                            Sentry
                          </Text>
                          <Text
                            accessibilityRole="text"
                            style={
                              {
                                "color": "#6a737d",
                                "fontFamily": "EuclidCircularB-Regular",
                                "fontSize": 14,
                                "fontWeight": "400",
                                "letterSpacing": 0,
                                "lineHeight": 22,
                                "marginTop": 8,
                              }
                            }
                          >
                            Generate a Developer Test Sentry trace.
                          </Text>
                          <TouchableOpacity
                            accessibilityRole="button"
                            accessible={true}
                            activeOpacity={1}
                            onPress={[Function]}
                            onPressIn={[Function]}
                            onPressOut={[Function]}
                            style={
                              {
                                "alignItems": "center",
                                "alignSelf": "stretch",
                                "backgroundColor": "transparent",
                                "borderColor": "#0376c9",
                                "borderRadius": 24,
                                "borderWidth": 1,
                                "flexDirection": "row",
                                "height": 48,
                                "justifyContent": "center",
                                "marginTop": 16,
                                "paddingHorizontal": 16,
                              }
                            }
                          >
                            <Text
                              accessibilityRole="text"
                              style={
                                {
                                  "color": "#0376c9",
                                  "fontFamily": "EuclidCircularB-Regular",
                                  "fontSize": 14,
                                  "fontWeight": "400",
                                  "letterSpacing": 0,
                                  "lineHeight": 22,
                                }
                              }
                            >
                              Generate Trace Test
                            </Text>
                          </TouchableOpacity>
                        </View>
                      </RCTScrollView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
